io_scene_godot/converters/material_node_tree/shader_functions.py

```diff
@@ -0,0 +1,621 @@
+"""Prewritten shader scripts for node in material node tree"""
+import re
+from ...structures import ValidationError
+
+FUNCTION_HEAD_PATTERN = re.compile(
+    (r'void\s+([a-zA-Z]\w*)\s*\(((\s*(out\s+)?'
+     r'(vec2|vec3|vec4|float|mat4|sampler2D)\s+[a-zA-Z]\w*\s*,?)*)\)'),
+)
+
+
+class ShaderFunction:
+    """Shader function for a blender node"""
+    def __init__(self, code):
+        # at most one group
+        self.code = code
+        self.in_param_types = list()
+        self.out_param_types = list()
+
+        matched_group = FUNCTION_HEAD_PATTERN.findall(code)[0]
+        self.name = matched_group[0]
+        parameters_str = matched_group[1]
+
+        for param_str in parameters_str.strip().split(','):
+            tokens = tuple([x.strip() for x in param_str.split()])
+            if tokens[0] == 'out':
+                self.out_param_types.append(tokens[1])
+            else:
+                self.in_param_types.append(tokens[0])
+
+
+class BsdfShaderFunction(ShaderFunction):
+    """Function for bsdf shader node, has additional information of
+    input and output socket"""
+    def __init__(self, code, input_sockets, output_sockets):
+        super().__init__(code)
+        # linked socket ids of material node
+        self.in_sockets = tuple(input_sockets)
+        self.out_sockets = tuple(output_sockets)
+
+
+'''
+Shader function nameing convention:
+
+The approach adopted by this addon to export material node is to
+convert it to a shader function. In order to simplify the mapping
+from Blender node to shader function, there is a naming convention
+for all the node functions.
+
+Blender node all have a entry `bl_idname` which is in the format
+of `ShaderNodexxx`. For an arbitrary Blender node, we drop the prefix
+in its `bl_idname` and convert it to snake case as the corresponding
+function name.
+
+    e.g. Blender node 'ShaderNodeBsdfPrincipled',
+            it's function name is node_bsdf_principled
+
+You may notice some Blender node has options like `space`, `operation`.
+In that case, if we want to convert node with differnet options to
+different functions, we are append the lowercased option string to the
+end of a function name.
+
+    e.g. Blender node 'ShaderNodeMath' with `operation` set as 'ADD',
+    `clamp` not checked. It's function name is 'node_math_add_no_clamp'.
+
+'''  # pylint: disable=W0105
```

Comment:
If you change this to a comment block, then you won't need to disable the pylint warning.

Otherwise, this comment may make sense moving it up to the module docstring.